tests/psa-client-server/psasim/include/client.h

```diff
@@ -1,5 +1,3 @@
-/* PSA Firmware Framework client header for psasim. */
-
 /*
```

Comment:
Part of the reason we're using psasim for client-server tests is that it implements the functional part of the [PSA firmware framework](https://developer.arm.com/documentation/den0063/latest/) specification. This means that the glue code to turn crypto function calls into remote procedure calls could, in principle, work on top of other PSA firmware framework implementations, and we could potentially productize it in the future.

This pull request removes or changes many standard macros and functions such as `PSA_FRAMEWORK_VERSION`, `psa_connect`, `psa_call`, etc. Is that really necessary?

Reply:
What I don't understand is how changing the communication _API_ and removing unrelated definitions helps with changing the communication method.

Reply:
A whole lot of things that weren't necessary were removed, thereby simplifying the code. That made changing the communication method easier. Why would we want to maintain parts of the code that aren't necessary?

Reply:
To be clear, I don't object to simplifying the implementation of the simulator by removing code, so that it would not do anything useful except run our tests. What I object to is changing the _interface_ of the simulator, which means that our glue code is no longer talking to a PSA framework implementation, but to some custom communication implementation. This reduces the confidence we could have that our code would work with a real-world communication interface.

For example, if `psa_connect` only supports a single SID and a single version, I want to keep the prototype `psa_handle_t psa_connect(uint32_t sid, uint32_t version)`. I don't mind if the implemenation is
```
if (version != 0) return 0;
if (sid != 42) return 0;
global_fd = open(...);
if (global_fd < 0) return 0;
return 1;
```

Reply:
As of 8111ab70dea9d544330c533ba8bb06b9e619b368 I'm happy that we're close enough to the official PSA framework API that when we build our client-server code on top of it, it's a good enough test of whether it could work on a real-world IPC implementation.

I'd prefer to have fewer modifications to the header files, so that we don't risk diverging more in the future. But for now we haven't diverged too much in my opinion.

Reply:
I agree with Gilles that ideally being as close as possible with the PSA FF interfaces in terms of the client-server comms is desirable for end users (with end users here I mean projects that consume this config). So I would as well recommend, if possible, to minimize the alterations made to the headers / prototypes. One possible use case of this is indeed that others project which are more strict with FF compliance will use these test configs to speed up validation without resorting to going on actual HW (or models of the HW, which might not be available upstream), so having less moving parts, e.g. get to compile with different headers from real world, as possible will be beneficial.

As a side note, I understand you're not serializing / deserializing as TF-M does, and this is not a problem, but did you evaluate/compare against how TF-M does it? In that case, if you can share any analysis it would be great for me to understand if we can improve on TF-M side.

Reply:
> One possible use case of this is indeed that others project which are more strict with FF compliance will use these test configs to speed up validation without resorting to going on actual HW

If with "configs" here you meant the Mbed TLS ones, then they have no impact on the client/server communication. Setting up the proper configs for Mbed TLS libraries is done in `all.sh` as for all other test components; `psasim` only implements the communication between client and server. So replacing `psasim` with any other implementation should work just fine.

> So I would as well recommend, if possible, to minimize the alterations made to the headers / prototypes

As far as I understood (and seen for example in Zephyr) anyone who provides a server implementation for PSA crypto also provides the "interface" counterpart to be used on the client side. This is done to make the 2 parties to work together both from the communication interface point of view (IPC vs trustzone vs something else) as well as from the serialization/deserialization one.

> As a side note, I understand you're not serializing / deserializing as TF-M does, and this is not a problem, but did you evaluate/compare against how TF-M does it?

Yes, when we started working on this we took a look at how the (de)serialization is done in TF-M in Zephyr, but it seemed to us that in that case files were manually created[^1] whereas we were looking for something automatically generated by a script. @tom-cosgrove-arm took care of this part, so I only roughly know how it's working, but there is no intentional improvement against the TF-M solution. It's just a different implementation where client/server interface functions for PSA are automatically generated starting from "crypto.h".
If we don't like this solution, I think that in the future we can migrate `psasim` to use the same (de)serialization pattern as TF-M, but IMO this is definitely out of scope for this PR.

### Final remark

It seems that with this PR I raised a lot of concerns so I would like to spend a few words on it.
This PR belongs to an EPIC whose goal was to prove that Mbed TLS can be built and tested in a "pure crypto client" scenario (i.e. `CRYPTO_CLIENT && !CRYPTO_C`). Up to 3.6 release there was no official statement for this to be supported, whereas now we can claim that "pure crypto client" can be built and tested like any other configuration used in test components. So I think the goal is (or will be - there are still couple of PRs ongoing) achieved.
`psasim` is where we started to get to this point, but:

- after playing a bit with it in the initial PRs we realized that it was not really tested (we fixed some bugs) so my conclusion was: if it has not been extensively tested from a functional behavior point of view, how can it claim that it's PSA-FF compliant?
- as I said above, in my understanding the goal of the EPIC was to prove that a "pure crypto client" configuration was possible in Mbed TLS, no matter how the communication between the 2 parties was implemented.

That's why I thought that in the end heavily reshaping it could not be too bad.
Should this be the final state then? Not necessarily. If in the future we find/design a real/good PSA-FF implementation, then we can easily replace `psasim` with that one[^2] and run tests in the same way. But IMO this is out of scope for this PR (and perhaps EPIC).


[^1]: of course please apologize if this is not correct, but none of us was an expert in TF-M so we might have overlooked features/behaviors there.
[^2]: replacing both client and server interfaces, of course.